lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,28 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
+      end
+
+      # @private
+      class Memoized
+        def initialize
+          @memoized = {}
+          @mutex = ReentrantMutex.new
+        end
+
+        def fetch_or_store(key)
+          @mutex.synchronize do
+            @memoized.fetch(key) { @memoized[key] = yield }
+          end
+        end
```

Comment:
I'm realizing, though, that there may be a possibility of hitting deadlock if dependent `let` definitions are written in a particular way and are being accessed by multiple threads. Not sure how likely that is to occur in practice (it may require someone actively trying to hit that case by coordinating threads via a `Queue`) but figured I'd mention it.